jwt.go

```diff
@@ -179,18 +182,72 @@ func (ja *JWTAuth) usingJWK() bool {
 }
 
 func (ja *JWTAuth) setupJWKLoader() {
+	// Initialiser le cache pour toutes les URL possibles
+	ja.jwkCaches = make(map[string]*struct {
+		cache     *jwk.Cache
+		cachedSet jwk.Set
+	})
+	ja.logger.Info("JWK cache initialized for placeholder URL", zap.String("placeholder_url", ja.JWKURL))
+}
+
+// getOrCreateJWKCache retrieves or creates a cache for a specific JWK URL
+func (ja *JWTAuth) getOrCreateJWKCache(resolvedURL string) (*struct {
+	cache     *jwk.Cache
+	cachedSet jwk.Set
+}, error) {
+	// If the URL is empty, return an error
+	if resolvedURL == "" {
+		return nil, fmt.Errorf("resolved JWK URL is empty")
+	}
+
+	// Check if cache already exists for this URL
+	if entry, ok := ja.jwkCaches[resolvedURL]; ok {
+		return entry, nil
+	}
+
+	// Create a new cache for this URL
 	cache := jwk.NewCache(context.Background(), jwk.WithErrSink(ja))
-	cache.Register(ja.JWKURL)
-	ja.jwkCache = cache
-	ja.refreshJWKCache()
-	ja.jwkCachedSet = jwk.NewCachedSet(cache, ja.JWKURL)
-	ja.logger.Info("using JWKs from URL", zap.String("url", ja.JWKURL), zap.Int("loaded_keys", ja.jwkCachedSet.Len()))
+	err := cache.Register(resolvedURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register JWK URL: %w", err)
+	}
+
+	// Try to refresh the cache immediately
+	_, err = cache.Refresh(context.Background(), resolvedURL)
+	if err != nil {
+		ja.logger.Warn("failed to refresh JWK cache during initialization", zap.Error(err), zap.String("url", resolvedURL))
+		// Continue even in case of error, the important thing is that the URL is registered
+	}
```

Comment:
This block is a repeat of `refreshJWKCache(resolvedURL)`. Creating fine grained functions can avoid this. For example:

```go
refreshJWKCache(entry, resolvedURL)

// or even better when we have the new type for the entry
entry.refreshCache()
```